frontend/src/pages/NewRun.test.tsx

```diff
@@ -0,0 +1,1160 @@
+    const experiment = newMockExperiment();
+    const props = generateProps();
+    props.location.search =
+      `?${QUERY_PARAMS.experimentId}=${experiment.id}`
+      + `&${QUERY_PARAMS.firstRunInExperiment}=1`;
+
+    getExperimentSpy.mockImplementation(() => experiment);
+
+    const tree = shallow(<TestNewRun {...props as any} />);
+    await TestUtils.flushPromises();
+
+    expect(tree).toMatchSnapshot();
+    tree.unmount();
+  });
+
+  it('shows a page error if getExperiment fails', async () => {
+    // Don't actually log to console.
+    // tslint:disable-next-line:no-console
+    console.error = jest.spyOn(console, 'error').mockImplementationOnce(() => null);
+
+    const experiment = newMockExperiment();
+    const props = generateProps();
+    props.location.search = `?${QUERY_PARAMS.experimentId}=${experiment.id}`;
+
+    TestUtils.makeErrorResponseOnce(getExperimentSpy, 'test error message');
+
+    const tree = shallow(<TestNewRun {...props as any} />);
+    await TestUtils.flushPromises();
+
+    expect(updateBannerSpy).toHaveBeenLastCalledWith(expect.objectContaining({
+      additionalInfo: 'test error message',
+      message: `Error: failed to retrieve associated experiment: ${experiment.id}. Click Details for more information.`,
+      mode: 'error',
+    }));
+    tree.unmount();
+  });
+
+  it('fetches the associated pipeline if one is present in the query params', async () => {
+    const pipeline = newMockPipeline();
+    const props = generateProps();
+    props.location.search = `?${QUERY_PARAMS.pipelineId}=${pipeline.id}`;
+
+    getPipelineSpy.mockImplementation(() => pipeline);
+
+    const tree = shallow(<TestNewRun {...props as any} />);
+    await TestUtils.flushPromises();
+
+    expect(tree.state()).toHaveProperty('pipeline', pipeline);
+    expect(tree.state()).toHaveProperty('pipelineName', pipeline.name);
+    expect(tree.state()).toHaveProperty('errorMessage', 'Run name is required');
+    expect(tree).toMatchSnapshot();
+    tree.unmount();
+  });
+
+  it('shows a page error if getPipeline fails', async () => {
+    // Don't actually log to console.
+    // tslint:disable-next-line:no-console
+    console.error = jest.spyOn(console, 'error').mockImplementationOnce(() => null);
+
+    const pipeline = newMockPipeline();
+    const props = generateProps();
+    props.location.search = `?${QUERY_PARAMS.pipelineId}=${pipeline.id}`;
+
+    TestUtils.makeErrorResponseOnce(getPipelineSpy, 'test error message');
+
+    const tree = shallow(<TestNewRun {...props as any} />);
+    await TestUtils.flushPromises();
+
+    expect(updateBannerSpy).toHaveBeenLastCalledWith(expect.objectContaining({
+      additionalInfo: 'test error message',
+      message: `Error: failed to retrieve pipeline: ${pipeline.id}. Click Details for more information.`,
+      mode: 'error',
+    }));
+    tree.unmount();
+  });
+
+  describe('choosing a pipeline', () => {
+
+    beforeEach(() => {
```

Comment:
nit: Seems like this can just be done once? Maybe change to `beforeAll` or just move the mock implementation up top, since it's using `mockImplementation` anyway?